File: python/dec_02/dec_02.py
with open("input_02.txt", "r") as input:
    data = input.read().strip()
result1 = 0
result2 = 0

# iterate through ranges
for line in data.split(","):
    begin, end = line.split("-")
    # iterate through range
    for i in range(int(begin), int(end) + 1):
        token = str(i)
        #if even
        if len(token) % 2 == 0:
            mid = len(token) // 2
            # if symetrical
            if token[:mid] == token[mid:]:
                result1 += i
        # check for repeated pattern
        for pattern_len in range(1, len(token) // 2 + 1):
          # if divisible  
            if len(token) % pattern_len == 0:
                # check if token is a repeated pattern
                if token == token[0:pattern_len] * (len(token) // pattern_len):
                    result2 += i
                    break

print(f"first: {result1}")
print(f"second: {result2}")

File: ruby/dec_02/dec_02.rb
data = File.read("input_02.txt").strip
result1 = 0
result2 = 0

# iterate through ranges
data.split(",").each do |line|
  begin_s, end_s = line.split("-")
  (begin_s.to_i..end_s.to_i).each do |i|
    token = i.to_s
    # if even
    if token.length.even?
      mid = token.length / 2
      # if symmetrical
      if token[0...mid] == token[mid..-1]
        result1 += i
      end
    end
    # check for repeated pattern
    (1..(token.length / 2)).each do |pattern_len|
      if token.length % pattern_len == 0
        pattern = token[0, pattern_len]
        if pattern * (token.length / pattern_len) == token
          result2 += i
          break
        end
      end
    end
  end
end

puts "first: #{result1}"
puts "second: #{result2}"
